src/util/text.ts

```diff
@@ -433,22 +433,29 @@ function lookupInUnicodeMap(code: u16, map: u16[]): bool {
   return false;
 }
 
+/** Creates an indentation matching the number of specified levels. */
 const indentX1 = "  ";
 const indentX2 = "    ";
+const indentX3 = "      ";
 const indentX4 = "        ";
+const indentCache = new Map<i32,string>();
 
-/** Creates an indentation matching the number of specified levels. */
 export function indent(sb: string[], level: i32): void {
-  while (level >= 4) {
-    sb.push(indentX4);
-    level -= 4;
-  }
-  if (level >= 2) {
-    sb.push(indentX2);
-    level -= 2;
-  }
-  if (level) {
-    sb.push(indentX1);
+  if (level <= 4) {
+    switch (level) {
+      case 1: sb.push(indentX1); break;
+      case 2: sb.push(indentX2); break;
+      case 3: sb.push(indentX3); break;
+      case 4: sb.push(indentX4); break;
+    }
+  } else {
+    let indents: string;
+    if (indentCache.has(level)) {
+      indents = assert(indentCache.get(level));
+    } else {
+      indentCache.set(level, (indents = indentX1.repeat(level)));
+    }
+    sb.push(indents);
```

Comment:
Hmm, the contents of `indentCache` will linger around forever in the global here, and a lookup from the map may be more expensive than pushing multiple `indentX4`s to an array. Is also more complex. Hence wondering, would it perhaps be preferable to add more `indexXY`s instead?

Reply:
> Hmm, the contents of indentCache will linger around forever in the global here, and a lookup from the map may be more expensive than pushing multiple

No, I specially benchmarked and Map is always faster on JS and AS side

Reply:
I see. Do you think that it would make sense to limit the maximum length in `indentCache` to some reasonable value? Can be double of what one would assume or so, it's more about having a limit at all.

Reply:
I absolutely do not see the point in this, especially when you consider that in theory we can not have more than 20-30 spaces. This function is mainly used in practice for dts сщвупуи and printing errors and etc

Reply:
Perhaps I'm too pessimistic here. Just seems that it would allow compiler bombing with indents, affecting a later (bigger) compile step somewhat covertly.

Reply:
Alright! I create fallback to ordinal `indentX1.repeat` without a cache if `indentCache.size() > 1024`. Wdyt?